### PR TITLE
Refactor `LibGit2.set_remote_url`

### PR DESCRIPTION
**Pull request by omus**
Added new functions `set_remote_fetch_url` and `set_remote_push_url` which allow you to specify the fetch/push remote URLs independently. Additionally, I switched `set_remote_url` from using a the keyword remote to using an optional argument.

Fixes https://github.com/JuliaLang/julia/issues/21773

### CODE REVIEW

File: base/libgit2/libgit2.jl [hunk 1]
```diff
@@ -240,49 +240,6 @@ function is_ancestor_of(a::AbstractString, b::AbstractString, repo::GitRepo)
...
-    with(GitConfig, repo) do cfg
-        set!(cfg, "remote.$remote.url", url)
-
-        m = match(GITHUB_REGEX,url)
-        if m !== nothing
-            push = "git@github.com:$(m.captures[1]).git"
-            if push != url
-                set!(cfg, "remote.$remote.pushurl", push)
-            end
-        end
```
Comment (omus) [inline base/libgit2/libgit2.jl]: Caused #21773

**Comment by StefanKarpinski** [comment]
Feature or bugfix?

**Comment by tkelman** [comment]
looks like a little of both, plus a new deprecation

File: base/libgit2/remote.jl [hunk 2]
```diff
@@ -241,3 +265,69 @@ function push(rmt::GitRemote, refspecs::Vector{<:AbstractString};
...
+Set both the fetch and push `url` for `remote` for the git repository `repo`. The default
+name of the remote is "origin".
+
+# Examples
+
+```julia
+repo_path = joinpath("test_directory", "Example")
+repo = LibGit2.init(repo_path)
+url1 = "https://github.com/JuliaLang/Example.jl"
+LibGit2.set_remote_url(repo, url1, remote="upstream")
```
Comment (tkelman) [inline base/libgit2/remote.jl]: bit misleading to name the argument at the call site if it's positional

File: base/deprecated.jl [hunk 3]
```diff
@@ -1348,6 +1348,18 @@ end
 @deprecate srand(filename::AbstractString, n::Integer=4) srand(read!(filename, Array{UInt32}(Int(n))))
 @deprecate MersenneTwister(filename::AbstractString)  srand(MersenneTwister(0), read!(filename, Array{UInt32}(Int(4))))
 
+# when this deprecation is deleted, remove all calls to it, and all `remote=nothing` keyword
+# arguments, and rename `remote_arg` to `remote` in base/libgit2/remote.jl
+@eval LibGit2 function deprecate_remote_keyword(f, sig, remote_arg, remote_kwarg)
+    if remote_kwarg === nothing
+        return remote_arg
+    else
+        msg = "$f($sig; remote=\"$remote_kwarg\") is deprecated, use $f($sig, \"$remote_kwarg\")"
```
Comment (tkelman) [inline base/deprecated.jl]: these usually end with "instead"

File: test/libgit2.jl [hunk 4]
```diff
@@ -37,6 +37,15 @@ function with_libgit2_temp_home(f)
     end
 end
 
+function is_defined_remote(repo::LibGit2.GitRepo, remote::AbstractString)
+    try
+        LibGit2.get(LibGit2.GitRemote, repo, name)
```
Comment (tkelman) [inline test/libgit2.jl]: can this be done without try-catching?

File: test/libgit2.jl [hunk 5]
```diff
@@ -1073,6 +1082,47 @@ mktempdir() do dir
...
+    @testset "Modify remote" begin
+        repo = LibGit2.GitRepo(test_repo)
+        try
+            name = "test"
+            url = "https://test.com/repo"
+
+            @test !is_defined_remote(repo, name)
+
+            # Set just the fetch URL
+            LibGit2.set_remote_fetch_url(repo, url, name)
```
Comment (tkelman) [inline test/libgit2.jl]: this is testing the `repo::GitRepo` versions pretty well, and the existing block was testing some of the `path::AbstractString` versions, but not the ones you're newly adding

File: base/libgit2/remote.jl (hunk 2 above)
Comment (omus) [inline base/libgit2/remote.jl]: I missed this when I moved away from a keyword.

File: test/libgit2.jl (hunk 5 above)
Comment (omus) [inline test/libgit2.jl]: I'll extend the tests to cover those.

File: base/deprecated.jl (hunk 3 above)
Comment (omus) [inline base/deprecated.jl]: Ok. I modeled this off of the `deprecate_negate` function. Does the rest of this seem reasonable?

File: test/libgit2.jl (hunk 4 above)
Comment (omus) [inline test/libgit2.jl]: I'll look into this again tomorrow. I didn't find a LibGit2 method which seemed to cover this.

**Comment by omus** [comment]
> Feature or bugfix?

Definitely both. When addressing the bug I noticed that the GitConfig was being used to modify the remote rather than using the available LibGit2 functions. Using the config method is bad as it doesn't verify that things like the remote name is valid. After getting my hands into the code I noticed that the remote keyword was unnecessary which is how I ended up here.

**Comment by omus** [comment]
I'll add some docstrings to the new functions in addition to the review change requests.

File: base/deprecated.jl (hunk 3 above)
Comment (tkelman) [inline base/deprecated.jl]: I wrote that one, and went through a couple refactors of it. it does say instead in it.

the argument order differs from the libgit2 c functions', but I guess that's okay. positionals always run the risk of "what order is it supposed to be" when writing or reading code that uses the method, so I'm neutral on this part of the change

Comment (omus) [inline base/deprecated.jl]: > I modeled this off of the deprecate_negate function. Does the rest of this seem reasonable?

I should have been more specific. I was wondering if the `deprecate_remote_keyword` function was reasonable.

Comment (omus) [inline base/deprecated.jl]: > the argument order differs from the libgit2 c functions'

Re-ordering the functions to be `func(repo::GitRepo, remote::AbstractString, url::AbstractString)` would more closely match both LibGit2 and the git command line. Personally I am fine with always having to specify the remote name.

**Comment by StefanKarpinski** [comment]
I guess what I was getting at was is this backportable to 0.6 or does it need to stay on master?

**Comment by omus** [comment]
> I guess what I was getting at was is this backportable to 0.6 or does it need to stay on master?

I can separate the bugfix so it can be backported to 0.6.

File: base/libgit2/remote.jl (hunk 2 above)
Comment (omus) [inline base/libgit2/remote.jl]: Fixed

File: base/deprecated.jl (hunk 3 above)
Comment (omus) [inline base/deprecated.jl]: Done

File: test/libgit2.jl (hunk 4 above)
Comment (omus) [inline test/libgit2.jl]: Made `lookup_remote` to address this

File: test/libgit2.jl (hunk 5 above)
Comment (omus) [inline test/libgit2.jl]: Done

**Comment by omus** [comment]
A note for the future @tkelman: Only backport commit ed263e8

**Comment by omus** [comment]
Anything left to do here? Maybe a NEWS entry?

File: base/libgit2/remote.jl [hunk 6]
```diff
@@ -240,4 +285,89 @@ function push(rmt::GitRemote, refspecs::Vector{<:AbstractString};
...
+    set_remote_push_url(repo::GitRepo, remote_name, url)
+    set_remote_push_url(path::String, remote_name, url)
+
+Set the push `url` for the specified `remote_name` for the GitRepo or the git repository
+located at `path`. Typically git repos use "origin" as the remote name.
+"""
+function set_remote_push_url end
+
+function set_remote_push_url(repo::GitRepo, remote_name::AbstractString, url::AbstractString)
+@check ccall((:git_remote_set_pushurl, :libgit2), Cint,
```
Comment (tkelman) [inline base/libgit2/remote.jl]: under-indented

Comment (omus) [inline base/libgit2/remote.jl]: Nice catch

**Comment by omus** [comment]
Rebased to resolve conflict in "base/deprecated.jl"

**Comment by omus** [comment]
Ready for merge

**Comment by tkelman** [comment]
didn't add a news entry, were you going to?

**Comment by omus** [comment]
I wasn't planning on adding one but I will now.

**Comment by omus** [comment]
NEWS entry added.

**Pull request merged**